src/main/java/software/amazon/awssdk/crt/CrtResource.java

```diff
@@ -16,96 +16,291 @@
+
+        if (debugNativeObjects) {
+            Log.log(Log.LogLevel.Trace, String.format("Closing instance of class %s with %d remaining refs", this.getClass().getCanonicalName(), remainingRefs));
+        }
+
+        if (remainingRefs > 0) {
+            return;
         }
 
-        while(ownedSubResources.size() > 0) {
-            CrtResource r = ownedSubResources.pop();
+        release();
+
+        if (canReleaseReferencesImmediately()) {
+            releaseReferences();
+        }
+    }
+
+    /**
+     * Decrements the ref counts for all resources referenced by this resource.  Most resources will have this called
+     * from their close() function, but resources with asynchronous shutdown processes must have it called from a
+     * shutdown completion callback.
+     */
+    protected void releaseReferences() {
+        if (debugNativeObjects) {
+            Log.log(Log.LogLevel.Trace, String.format("Instance of class %s closing referenced objects", this.getClass().getCanonicalName()));
+        }
+
+        while(referencedResources.size() > 0) {
+            CrtResource r = referencedResources.pop();
             r.close();
         }
     }
+
+    /**
+     * Debug method to log all of the currently un-closed CRTResource objects.
+     */
+    public static void logNativeResources() {
+        Log.log(Log.LogLevel.Trace, "Dumping native object set:");
+        for (Map.Entry<Long, String> entry : NATIVE_RESOURCES.entrySet()) {
+            Log.log(Log.LogLevel.Trace, String.format(" * %s class instance using native pointer %d", entry.getValue(), entry.getKey().longValue()));
+        }
+    }
+
+    /**
+     * Debug method to increment the current native object count.
+     */
+    private static void incrementNativeObjectCount() {
+        if (!debugNativeObjects) {
+            return;
+        }
+
+        lock.lock();
+        try {
+            ++resourceCount;
+            Log.log(Log.LogLevel.Trace, String.format("incrementNativeObjectCount - count = %d", resourceCount));
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Debug method to decrement the current native object count.
+     */
+    private static void decrementNativeObjectCount() {
+        if (!debugNativeObjects) {
+            return;
+        }
+
+        lock.lock();
+        try {
+            --resourceCount;
+            Log.log(Log.LogLevel.Trace, String.format("decrementNativeObjectCount - count = %d", resourceCount));
+            if (resourceCount == 0) {
+                emptyResources.signal();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Debug/test method to wait for the CRTResource count to drop to zero.  Times out with an exception after
+     * a period of waiting.
+     */
+    public static void waitForNoResources() {
+        if (!debugNativeObjects) {
+            return;
+        }
+
+        lock.lock();
```

Comment:
How is this working?? By locking the `resourceCount` lock here, won't that stops other Threads from being able to decrement the int to zero?

Reply:
The lock is released during await.